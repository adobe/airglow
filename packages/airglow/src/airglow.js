/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';
import React from 'react';
import { AirglowProvider, AirglowWrapper } from './airglow.context';
import Engine from './engine';
import { COMPONENT, HOC } from './types';

class Airglow extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.store = this.props.store ? this.props.store() : null;
    this.engine = new Engine(this.store, R.reverse(this.props.plugins || []));

    if (this.props.bootstrap) {
      this.engine.plugin(
        HOC,
        { component: AirglowWrapper, props: { config: this.props.bootstrap } }
      );
    }
  }

  renderChildren() {
    // take our children and merge them with the render plugins
    return R.reduce(
      (children, plugin) =>
        pluginReducers[plugin.type](plugin.callback, children),
      this.props.children,
      R.reverse(getRenderPlugins(this.engine))
    );
  }

  render() {
    const Renderer = this.props.renderer || AirglowProvider;
    return (
      <Renderer
        value={{
          bootstrap: (...args) => this.engine.bootstrapModule(...args),
          dispatch: action => this.engine.getStore().dispatch(action)
        }}
      >
        {this.renderChildren()}
      </Renderer>
    );
  }
}

const reduceHoc = ({ component: Component, props }, children) =>
  <Component {...props}>{children}</Component>;

const reduceComponent = (component, children) => (
  <div id="airglowRootWrapper">
    {children}
    {component}
  </div>
);

const pluginReducers = {
  [HOC]: reduceHoc,
  [COMPONENT]: reduceComponent
};

const getRenderPlugins = engine => R.reverse(R.sortBy(
  R.prop('index'),
  R.concat(
    engine.getPluginsData(HOC) || [],
    engine.getPluginsData(COMPONENT) || []
  )
));

export default Airglow;
