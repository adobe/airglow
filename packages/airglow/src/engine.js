/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';
import { FEATURE, BOOTSTRAP_MODULE, BOOTSTRAP_PLUGIN } from './types';

let PLUGIN_INDEX = 0;

class Engine {
  constructor(storeFactory, plugins = []) {
    this.registery = {};
    this.bootstrapped = {};
    plugins.forEach(plugin => plugin(this));
    if (storeFactory) {
      this.store = this.generateStore(storeFactory);
    }
    this.enableFeatures();
    this.bootstrapPlugins();
  }

  plugin(type, callback) {
    this.registery[type] = (this.registery[type] || []).concat({
      type,
      callback,
      index: ++PLUGIN_INDEX
    });
  }

  getPlugins(type) {
    return R.map(R.prop('callback'), this.registery[type] || []);
  }

  getPluginsData(plugin) {
    return this.registery[plugin];
  }

  generateStore(storeFactory) {
    const store = storeFactory.create(type => this.getPlugins(type));
    storeFactory.bootstrap(this);
    return store;
  }

  getStore() {
    return this.store;
  }

  enableFeatures() {
    this.getPlugins(FEATURE).forEach(feature =>
      feature(type => this.getPlugins(type)));
  }

  bootstrapPlugins() {
    this.getPlugins(BOOTSTRAP_PLUGIN).forEach(
      bootstrapper => bootstrapper({ dispatch: this.store ? this.store.dispatch : null })
    );
  }

  bootstrapModule({ name, reducers, ...options }) {
    if (this.bootstrapped[name]) { return; }
    if (this.store) {
      this.store.injectReducers(reducers);
    }

    const fullOptions = {
      dispatch: this.store ? this.store.dispatch : null,
      ...options
    };

    this.getPlugins(BOOTSTRAP_MODULE).forEach(
      bootstrapper => bootstrapper(fullOptions)
    );
    this.bootstrapped[name] = true;
  }
}

export default Engine;
