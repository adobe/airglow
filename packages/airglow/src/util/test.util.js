/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import React from 'react';
import { mount } from 'enzyme';
import Airglow from '../airglow';

export const renderAirglow = (children, props) => {
  let airglow;
  const App = () => {
    airglow = (
      <Airglow {...props}>
        {children}
      </Airglow>
    );
    return airglow;
  };

  const tree = mount(<App />);
  tree.airglow = tree;
  tree.store = tree.find('Airglow').instance().engine.getStore();
  if (tree.store) {
    tree.dispatch = tree.store.dispatch;
    tree.getState = tree.store.getState;
  }
  return tree;
};

class FakeStore {
  constructor() {
    this.bootstrap = sinon.spy();
  }

  create(...args) {
    this.createArgs = args;
    this.store = {
      dispatch: sinon.spy(),
      getState: sinon.spy(),
      injectReducers: sinon.spy()
    };
    return this.store;
  }
}

export const fakeStore = () => new FakeStore();
