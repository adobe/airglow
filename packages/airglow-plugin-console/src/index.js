/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';
import { REDUCER, BOOTSTRAP_MODULE, BOOTSTRAP_PLUGIN } from 'airglow';
import { slice, createAction } from '@airglow/reducers';
import { BOOTSTRAP_CONSOLE } from './actions';
import reducer from './reducers';

const STORE_KEY = 'console-commands';
const doBootstrap  = createAction(BOOTSTRAP_CONSOLE);

export default function prefabPlugin(config) {
  return (engine) => {
    engine.plugin(REDUCER, slice(STORE_KEY).with(reducer));
    engine.plugin(BOOTSTRAP_MODULE, bootstrap(engine));
    engine.plugin(BOOTSTRAP_PLUGIN, bootstrapPlugin(engine, config));
  };
}

const logHelp = ({ state }) => {
  R.forEachObjIndexed(
    ({ description }, key) => {
      console.info(`${key}(): ${description}`); // eslint-disable-line no-console
    },
    state['console-commands']
  );
};

const bootstrapPlugin = (
  engine, { help = 'help', helpDescription = 'Display this help menu', consoleCommands }
) => ({ dispatch }) => {
  bootstrap(engine)({
    dispatch,
    consoleCommands: {
      [help]: { description: helpDescription, command: logHelp },
      ...consoleCommands
    }
  });
};

const bootstrap = engine => ({ dispatch, consoleCommands }) => {
  if (consoleCommands) {
    dispatch(doBootstrap(consoleCommands));
    R.forEachObjIndexed(
      ({ command }, key) => {
        window[key] = () => command({ state: engine.getStore().getState(), dispatch });
      },
      consoleCommands
    );
  }
};
