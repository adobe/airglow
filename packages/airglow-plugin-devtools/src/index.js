/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import React from 'react';
import { POST_ENHANCER, COMPONENT, COMPOSER } from 'airglow';
import Devtools from './devtools';

export default function devtoolsPlugin({
  reduxDevtools = false,
  reduxBrowser = false
} = {}) {
  return (engine) => {
    if (reduxDevtools) {
      engine.plugin(POST_ENHANCER, Devtools.instrument());
      engine.plugin(COMPONENT, <Devtools key="c" />);
    }
    if (reduxBrowser) {
      engine.plugin(COMPOSER, browserCompose);
    }
  };
}

/* eslint-disable no-underscore-dangle */
const browserHook = () => global.window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__;

const browserCompose = compose =>
  (browserHook() ? browserHook()({}) : compose);
