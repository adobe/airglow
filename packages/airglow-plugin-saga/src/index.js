/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import createSagaMiddleware from 'redux-saga';
import { FEATURE, MIDDLEWARE, BOOTSTRAP_MODULE } from 'airglow';

export const SAGA = 'wire_sagas';

export default function sagaPlugin() {
  const middleware = createSagaMiddleware();

  return (engine) => {
    engine.plugin(MIDDLEWARE, middleware);
    engine.plugin(FEATURE, sagaFeature(middleware));
    engine.plugin(BOOTSTRAP_MODULE, bootstrapSagas(middleware));
  };
}

const sagaFeature = middleware => lookup =>
  lookup(SAGA).forEach(
    plugin => middleware.run(plugin)
  );

const bootstrapSagas = middleware => ({ sagas }) => {
  if (!sagas) { return; }
  const runSagas = Array.isArray(sagas) ? sagas : [sagas];
  runSagas.forEach(saga => middleware.run(saga));
};
