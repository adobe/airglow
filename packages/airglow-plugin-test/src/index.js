/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { REDUCER } from 'airglow';
import { slice } from '@airglow/reducers';
import { last } from 'ramda';
import reducer from './reducer';

const STORE_KEY = '_test_dispatches';

export default function testPlugin() {
  return (engine) => {
    engine.plugin(REDUCER, slice(STORE_KEY).with(reducer));
  };
}

export const lastDispatch = state => last(state[STORE_KEY]);
export const dispatchList = state => state[STORE_KEY];
