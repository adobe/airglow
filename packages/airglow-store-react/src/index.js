/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';

import { REDUCER, HOC } from 'airglow';
import { StoreProvider } from './store.context';

export { default as connect } from './connect';

let SUBSCRIBE_INDEX = 0;

class Store {
  constructor(reducers = []) {
    this.state = {};
    this.reducers = reducers;
    this.dispatch = this.dispatch.bind(this);
    this.injectReducers = this.injectReducers.bind(this);
    this.subscriptions = {};
    this.dispatch({ type: 'REACT.STORE.INITIALIZE' });
  }

  dispatch(action) {
    this.state = R.reduce(
      (newState, reducer) => reducer(newState, action),
      this.state,
      this.reducers || []
    );
    for (let i = 0; i <= SUBSCRIBE_INDEX; ++i) {
      if (this.subscriptions[i]) {
        this.subscriptions[i](this);
      }
    }
  }

  getState() {
    return this.state;
  }

  subscribe(callback) {
    this.subscriptions[++SUBSCRIBE_INDEX] = callback;
    return () => { this.subscriptions[SUBSCRIBE_INDEX] = null; };
  }

  injectReducers(newReducers) {
    if (!newReducers) { return; }
    this.reducers = R.concat(this.reducers, newReducers);
    this.dispatch({ type: 'REACT.STORE.INJECT.REDUCERS' });
  }
}

export const createStore = reducers => new Store(reducers);

class StoreFactory {
  create(lookup) {
    this.store = createStore(getReducers(lookup));
    return this.store;
  }

  bootstrap(engine) {
    engine.plugin(HOC, {
      component: StoreProvider,
      props: {
        store: this.store
      }
    });
  }
}

const getReducers = lookup => reduceLookup(lookup(REDUCER));

const reduceLookup = data => data.reduce(
  (results, plugin) => mergeInto(results, plugin), []
);

const mergeInto = (source, toMerge) =>
  source.concat(Array.isArray(toMerge) ? toMerge : [toMerge]);

export default () => new StoreFactory();
