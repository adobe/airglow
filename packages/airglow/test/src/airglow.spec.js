/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import React from 'react';
import sinon from 'sinon';
import { mount, shallow } from 'enzyme';

import Airglow from '../../src/airglow';
import * as engine from '../../src/engine';
import { COMPONENT, HOC } from '../../src/types';
import { fakeStore } from '../../src/util/test.util';

let bootstrapSpy;


const Renderer = props => (
  <div {...props} />
);

const components = [
  {
    index: 1,
    callback: <div key="V">Version 2.1</div>,
    type: COMPONENT
  },
  {
    index: 5,
    callback: <div key="Q">Smokey</div>,
    type: COMPONENT
  }
];
const wrappers = [
  {
    index: 3,
    callback: {
      component: props => <div {...props} />,
      props: { className: 'sugar', key: 'C' }
    },
    type: HOC
  }
];

describe('airglow', () => {
  afterEach(function () {
    engine.default.restore();
  });

  const stubEngine = ({ components: cs = [], wrappers: wrs = [] } = {}) => {
    bootstrapSpy = sinon.spy();
    sinon.stub(engine, 'default').callsFake(() => ({
      store: 'STORE!',
      getPluginsData: (type) => {
        if (type === COMPONENT) { return cs; }
        if (type === HOC) { return wrs; }
        return [];
      },
      bootstrapModule: bootstrapSpy
    }));
  };

  it('should pass the plugins into the engine', function () {
    stubEngine();
    shallow(<Airglow store={fakeStore} plugins={['a', 'b', 'c']} />);
    expect(engine.default.lastCall.args[0]).toMatchSnapshot();
  });

  it('should render children correctly', function () {
    stubEngine();
    expect(mount(
      <Airglow store={fakeStore} renderer={Renderer}>
        <button type="button">Test</button>
      </Airglow>
    ).text()).toBe('Test');
  });

  it('should accept multiple children', function () {
    stubEngine();
    const tree = mount(
      <Airglow store={fakeStore} renderer={Renderer}>
        <button type="button" key="TestA">Test A</button>
        <button type="button" key="TestB">Test B</button>
      </Airglow>
    );

    expect(tree.find('button')).toHaveLength(2);
  });

  it('should append component plugins', function () {
    stubEngine({ components });
    const tree = mount(
      <Airglow store={fakeStore} renderer={Renderer}>
        <div key="TestA">Test</div>
      </Airglow>
    );
    expect(tree.debug()).toMatchSnapshot();
  });

  it('should wrap higher order components around the content', function () {
    stubEngine({ wrappers });
    const tree = mount(
      <Airglow store={fakeStore} renderer={Renderer}>
        <div key="A">Test</div>
      </Airglow>
    );
    expect(tree.debug()).toMatchSnapshot();
  });

  it('should render in order', () => {
    stubEngine({ wrappers, components });
    const tree = mount(
      <Airglow store={fakeStore} renderer={Renderer}>
        <div key="A">Test</div>
      </Airglow>
    );
    expect(tree.debug()).toMatchSnapshot();
  });

  it('should use the engines bootstrap function', function () {
    stubEngine();
    const rendered = mount(<Airglow store={fakeStore} plugins={[]} renderer={Renderer} />);
    rendered.find(Renderer).props().value.bootstrap('zipper');
    expect(bootstrapSpy.calledWith('zipper')).toBe(true);
  });
});
