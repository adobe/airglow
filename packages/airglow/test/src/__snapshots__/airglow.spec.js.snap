// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`airglow creates an AirglowWrapper HOC if bootstrap is specified 1`] = `
Array [
  "higherOrderComponent",
  Object {
    "component": [Function],
    "props": Object {
      "config": "bootstrap data",
    },
  },
]
`;

exports[`airglow should append component plugins 1`] = `
"<Airglow store={[Function: fakeStore]} renderer={[Function: Renderer]}>
  <Renderer value={{...}}>
    <div value={{...}}>
      <div id=\\"airglowRootWrapper\\">
        <div id=\\"airglowRootWrapper\\">
          <div>
            Test
          </div>
          <div>
            Smokey
          </div>
        </div>
        <div>
          Version 2.1
        </div>
      </div>
    </div>
  </Renderer>
</Airglow>"
`;

exports[`airglow should pass the plugins into the engine 1`] = `
FakeStore {
  "bootstrap": [Function],
}
`;

exports[`airglow should render in order 1`] = `
"<Airglow store={[Function: fakeStore]} renderer={[Function: Renderer]}>
  <Renderer value={{...}}>
    <div value={{...}}>
      <div id=\\"airglowRootWrapper\\">
        <component className=\\"sugar\\">
          <div className=\\"sugar\\">
            <div id=\\"airglowRootWrapper\\">
              <div>
                Test
              </div>
              <div>
                Smokey
              </div>
            </div>
          </div>
        </component>
        <div>
          Version 2.1
        </div>
      </div>
    </div>
  </Renderer>
</Airglow>"
`;

exports[`airglow should wrap higher order components around the content 1`] = `
"<Airglow store={[Function: fakeStore]} renderer={[Function: Renderer]}>
  <Renderer value={{...}}>
    <div value={{...}}>
      <component className=\\"sugar\\">
        <div className=\\"sugar\\">
          <div>
            Test
          </div>
        </div>
      </component>
    </div>
  </Renderer>
</Airglow>"
`;
